src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,448 @@
+        private nint NextJSVHandle = -2;
+        private readonly List<nint> JSVHandleFreeList = new();
+
+#if !FEATURE_WASM_THREADS
+        public static readonly JSProxyContext DefaultInstance = new();
+        public static JSProxyContext CurrentInstance => DefaultInstance;
+        public static JSProxyContext CapturedInstance
+        {
+            get => MainInstance!;
+            set { }
+        }
+        public static JSProxyContext MainInstance => DefaultInstance;
+#else
+        public nint TargetTID;
+        public int ThreadId;
+        public bool IsMainThread;
+        public JSSynchronizationContext SynchronizationContext;
+
+        [UnsafeAccessor(UnsafeAccessorKind.Field, Name = "thread_id")]
+        private static extern ref long GetThreadNativeThreadId(Thread @this);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static IntPtr GetNativeThreadId()
+        {
+            return (int)GetThreadNativeThreadId(Thread.CurrentThread);
+        }
+
+        // Context of the main thread
+        private static JSProxyContext? _MainInstance;
+        public static JSProxyContext MainInstance
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _MainInstance!;
+            set => _MainInstance = value;
+        }
+
+        // Context captured from parameters of the current operation.
+        // Most of the time, this matches the current thread.
+        // When we pass `JSObject` instance to `JSImport`, it could have thread affinity different than the current thread.
+        // We will use that parameter to capture the target thread to call into.
+        [ThreadStatic]
+        public static JSProxyContext? CapturedInstance;
+
+        // Context of the current thread. Could be null on threads which don't have JS interop, like managed thread pool threads.
+        [ThreadStatic]
+        public static JSProxyContext? CurrentInstance;
+
+        // This is context to dispatch into. In order of preference
+        // - captured context by arguments of current/pending JSImport call
+        // - current thread context, for calls from JSWebWorker threads with the interop installed
+        // - main thread, for calls from any other thread, like managed thread pool or `new Thread`
+        public static JSProxyContext DefaultInstance
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                // until we set CapturedInstance in the generated code of JSImport/JSExport, we need to fallback to CurrentInstance to find target thread
+                // because maybe it was not captured by the other parameters yet
+                // after capture is solid, we could drop DefaultInstance and use CapturedInstance instead
+                // TODO: sort generated ToJS() calls to make the capture first
+                return CapturedInstance ?? CurrentInstance ?? MainInstance;
```

Comment:
this makes me really nervous. i'm used to this particular pattern going wrong, where we have this carefully nested set of states that cause operations to magically go to the right spot. it's fine for now probably but I would strongly prefer that we never have ?? chains like this when it comes to selecting the appropriate thread for an operation